### PR TITLE
[136172] Added rowsPerPageTitle to customize rows per page text 

### LINKED ISSUE #136172: PaginatedDataTable Pagination Footer Section Customizations
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

The current implementation of Footer Section in Pagination area of PaginatedDataTable works but this Class could be more useful if we can customize it a bit. 

I have few things in my mind which we can work on: 

1. Pagination Footer Text (Currently it defaults to 'Rows per page:') #136173
2. Pagination Footer Alignment (Currently it defaults to Right side) #137507

### Proposal

Currently:

Screenshot 2023-10-08 at 8 09 40 PM

Can be changed with custom text and custom center alignment to this : 

image

### PR DESCRIPTION
**Pull request by gitabhisingh**
Added rowsPerPageTitle to customize rows per page text for pagination footer. Fixes 136172

Fixes : https://github.com/flutter/flutter/issues/136172

Added  new parameter `rowsPerPageTitle` in order to customize the Rows Per Page text for Pagination in PaginatedDataTableClass

**Screenshots:**

Before: 
image

After: 
image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (TahaTesser) [inline packages/flutter/lib/src/material/paginated_data_table.dart:265]: `localizations.this.rowsPerPageTitle]` is an invalid reference.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Thanks for the PR.  Can you please add tests for this parameter?

**Comment by gitabhisingh** [comment]
Added unit test for the new Parameter. @TahaTesser

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (TahaTesser) [inline packages/flutter/test/material/paginated_data_table_test.dart:206]:
```suggestion
  testWidgetsWithLeakTracking('Default rowsPerPageTitle', (WidgetTester tester) async {
    await tester.pumpWidget(MaterialApp(
      home: PaginatedDataTable(
        header: const Text('Test table'),
        source: source,
        // This callback shows the rows page title in the footer.
        onRowsPerPageChanged: (int? rowsPerPage) { },
        rowsPerPage: 2,
        availableRowsPerPage: const <int>[
          2, 3, 4, 5, 7, 8,
        ],
        columns: const <DataColumn>[
          DataColumn(label: Text('Name')),
          DataColumn(label: Text('Calories'), numeric: true),
          DataColumn(label: Text('Generation')),
        ],
      ),
    ));

    final MaterialLocalizations localizations = MaterialLocalizations.of(
      tester.element(find.byType(PaginatedDataTable)),
    );
    expect(find.text(localizations.rowsPerPageTitle), findsOneWidget);
  });
```
Good call on adding default test.

This can be cleaned up  a little bit and test the `MaterialLocalizations` string.

File: packages/flutter/test/material/paginated_data_table_test.dart
Comment (TahaTesser) [inline packages/flutter/test/material/paginated_data_table_test.dart:239]:
```suggestion
  testWidgetsWithLeakTracking('PaginatedDataTable.rowsPerPageTitle overrides the default rowsPerPageTitle', (WidgetTester tester) async {
    Widget buildTable({String? rowsPerPageTitle}) {
      return MaterialApp(
        home: PaginatedDataTable(
          header: const Text('Test table'),
          source: source,
          rowsPerPageTitle: rowsPerPageTitle,
          // This callback shows the rows page title in the footer.
          onRowsPerPageChanged: (int? rowsPerPage) { },
          rowsPerPage: 2,
          availableRowsPerPage: const <int>[
            2, 3, 4, 5, 7, 8,
          ],
          columns: const <DataColumn>[
            DataColumn(label: Text('Name')),
            DataColumn(label: Text('Calories'), numeric: true),
            DataColumn(label: Text('Generation')),
          ],
        ),
      );
    }

    // Test default rowsPerPageTitle.
    await tester.pumpWidget(buildTable());

    MaterialLocalizations localizations = MaterialLocalizations.of(
      tester.element(find.byType(PaginatedDataTable)),
    );
    expect(find.text(localizations.rowsPerPageTitle), findsOneWidget);

    // Test custom rowsPerPageTitle.
    await tester.pumpWidget(buildTable(rowsPerPageTitle: 'Items per page:'));

    localizations = MaterialLocalizations.of(
      tester.element(find.byType(PaginatedDataTable)),
    );

    expect(find.text(localizations.rowsPerPageTitle), findsNothing);
    expect(find.text('Items per page:'), findsOneWidget);
  });
```

Similar to suggested default test.

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (TahaTesser) [inline packages/flutter/lib/src/material/paginated_data_table.dart:266]:
```suggestion
  /// When provided, this string replaces the rows per title in the footer.
  ///
  /// Defaults to [MaterialLocalizations.rowsPerPageTitle].
  final String? rowsPerPageTitle;
```
`localizations.rowsPerPageTitle` reference is still invalid as it is looking for `MaterialLocalizations` itself, not the variable.

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Great job on adding tests. 

I've made some improvements to land this.

**Comment by gitabhisingh** [comment]
Thanks for the suggestions @TahaTesser.
Updated branch with your suggestions and unit tests passing locally 🚀

**Review by TahaTesser — APPROVED** [review]
LGTM! 

We'll need a second review to land this.

**Comment by gitabhisingh** [comment]
Hey @Piinks requesting your review here :)

File: packages/flutter/lib/src/material/paginated_data_table.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/paginated_data_table.dart:485]: I think it might be best to approach customizing this through localizations. Looking at the rest of the footer items using localizations, if we expose this one, why not all of them? Enumerating so many properties on a widget when they can be accessed and overridden via an object like localizations is not typically a change we would make here.

**Review by Piinks — COMMENTED** [review]
Hi @gitabhisingh thanks for contributing! Have you tried overriding the localizations to provide the preferred text in this case?

**Comment by gitabhisingh** [comment]
Closing as we might want to use Localization to achieve this

**Pull request closed without merging**
